types/default.nix

```diff
@@ -0,0 +1,346 @@
+        ${../disk-deactivate}/disk-deactivate "$dev" | bash -x
+      done
+
+      echo 'creating partitions...'
+      ${diskoLib.create devices}
+      echo 'mounting partitions...'
+      ${diskoLib.mount devices}
+    '';
+    /* Takes a disko device specification and returns a nixos configuration
+
+       config :: lib.types.devices -> nixosConfig
+    */
+    config = devices: flatten (map (dev: dev._config) (flatten (map attrValues (attrValues devices))));
+    /* Takes a disko device specification and returns a function to get the needed packages to format/mount the disks
+
+       packages :: lib.types.devices -> pkgs -> [ derivation ]
+    */
+    packages = devices: pkgs: unique (flatten (map (dev: dev._pkgs pkgs) (flatten (map attrValues (attrValues devices)))));
+  };
+
+  optionTypes = rec {
+    # POSIX.1‐2017, 3.281 Portable Filename
+    filename = lib.mkOptionType {
+      name = "POSIX portable filename";
+      check = x: isString x && builtins.match "[0-9A-Za-z._][0-9A-Za-z._-]*" x != null;
+      merge = mergeOneOption;
+      description = "A filename";
+    };
+
+    # POSIX.1‐2017, 3.2 Absolute Pathname
+    absolute-pathname = lib.mkOptionType {
+      name = "POSIX absolute pathname";
+      check = x: isString x && substring 0 1 x == "/" && pathname.check x;
+      merge = mergeOneOption;
+      description = "An absolute path";
+    };
+
+    # POSIX.1-2017, 3.271 Pathname
+    pathname = lib.mkOptionType {
+      name = "POSIX pathname";
+      check = x:
+        let
+          # The filter is used to normalize paths, i.e. to remove duplicated and
+          # trailing slashes.  It also removes leading slashes, thus we have to
+          # check for "/" explicitly below.
+          xs = filter (s: stringLength s > 0) (splitString "/" x);
+        in
+        isString x && (x == "/" || (length xs > 0 && all filename.check xs));
+      merge = mergeOneOption;
+      description = "A path name";
+    };
+  };
+
+  /* topLevel type of the disko config, takes attrsets of disks, mdadms, zpools, nodevs, and lvm vgs.
+  */
+  devices = lib.types.submodule {
+    options = {
+      disk = lib.mkOption {
+        type = lib.types.attrsOf subTypes.disk;
+        default = { };
+        description = "Block device";
+      };
+      mdadm = lib.mkOption {
+        type = lib.types.attrsOf subTypes.mdadm;
+        default = { };
+        description = "mdadm device";
+      };
+      zpool = lib.mkOption {
+        type = lib.types.attrsOf subTypes.zpool;
+        default = { };
+        description = "ZFS pool device";
+      };
+      lvm_vg = lib.mkOption {
+        type = lib.types.attrsOf subTypes.lvm_vg;
+        default = { };
+        description = "LVM VG device";
+      };
+      nodev = lib.mkOption {
+        type = lib.types.attrsOf subTypes.nodev;
+        default = { };
+        description = "A non-block device";
+      };
+    };
+  };
+
+  subTypes = {
+    nodev = diskoLib.mkSubType [ ./nodev.nix ];
```

Comment:
why does this take a list?

Reply:
modules

Reply:
https://github.com/nix-community/disko/pull/131/files#diff-63b2750ca88f0bb1b20a657859ddd5df357da7d87777bfd4054db88de1a6a598R148

Reply:
but we don't need the list interface? it could be just a file?

Reply:
Done.